src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+namespace Microsoft.Rest
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Abstract base class for types that act like enums, but can be extended with arbitrary string values.
+    /// </summary>
+    public abstract class ExtensibleEnum<T> : IEquatable<T> where T : ExtensibleEnum<T>
+    {
+        private static readonly Lazy<Dictionary<string, T>> _nameMap =
+            new Lazy<Dictionary<string, T>>(CreateNameMap, isThreadSafe: true);
+
+        private string _name;
+
+        /// <summary>
+        /// Initializes a new instance of the ExtensibleEnum class.
+        /// </summary>
+        /// <param name="name">The value for this instance of the extensible enumeration.</param>
+        protected ExtensibleEnum(string name)
+        {
+            _name = name;
+        }
+
+        /// <summary>
+        /// Compares two ExtensibleEnum values for equality.
+        /// </summary>
+        /// <param name="lhs">The first ExtensibleEnum to compare.</param>
+        /// <param name="rhs">The second ExtensibleEnum to compare.</param>
+        /// <returns>true if the ExtensibleEnum objects are equal or are both null; false otherwise.</returns>
+        public static bool operator ==(ExtensibleEnum<T> lhs, T rhs) => object.Equals(lhs, rhs);
+
+        /// <summary>
+        /// Compares two ExtensibleEnum values for inequality.
+        /// </summary>
+        /// <param name="lhs">The first ExtensibleEnum to compare.</param>
+        /// <param name="rhs">The second ExtensibleEnum to compare.</param>
+        /// <returns>true if the ExtensibleEnum objects are not equal; false otherwise.</returns>
+        public static bool operator !=(ExtensibleEnum<T> lhs, T rhs) => !object.Equals(lhs, rhs);
+
+        /// <summary>
+        /// Defines explicit conversion from ExtensibleEnum to string.
+        /// </summary>
+        /// <param name="name">ExtensibleEnum to convert.</param>
+        /// <returns>The ExtensibleEnum as a string.</returns>
+        public static explicit operator string(ExtensibleEnum<T> name) => name?.ToString();
+
+        /// <summary>
+        /// Looks up an ExtensibleEnum instance by name, or returns null if the given name does not match one of the
+        /// known values of this ExtensibleEnum.
+        /// </summary>
+        /// <param name="name">Name of the ExtensibleEnum value.</param>
+        /// <returns>An instance of type T with the given name, or null if no such instance could be found.</returns>
+        protected static T Lookup(string name)
```

Comment:
perhaps we should change this to a static indexer -

``` csharp
protected static T this[string value] {
... 
}
```

Reply:
I'm pretty sure that static indexers are [not](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/indexers/) a [thing](https://stackoverflow.com/questions/401232/static-indexers).

Also, the intent of this method is to provide derived classes with a way to implement `Create` with interning semantics. Given that it's protected, I'm not sure what the extra syntactic sugar would buy you.

Reply:
Yup, learned this the hard way